src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs

```diff
@@ -99,6 +150,124 @@ public void OnDiagnosticsUnhandledException(HttpContext httpContext, Exception e
+
+            StringValues requestId;
+            StringValues standardRootId;
+            if (httpContext.Request.Headers.TryGetValue(RequestResponseHeaders.RequestIdHeader, out requestId))
+            {
+                activity.SetParentId(requestId);
+                requestTelemetry.Context.Operation.ParentId = requestId;
+
+                string[] baggage = httpContext.Request.Headers.GetCommaSeparatedValues(RequestResponseHeaders.CorrelationContextHeader);
+                if (baggage != StringValues.Empty)
+                {
+                    foreach (var item in baggage)
+                    {
+                        NameValueHeaderValue baggageItem;
+                        if (NameValueHeaderValue.TryParse(item, out baggageItem))
+                        {
+                            activity.AddBaggage(baggageItem.Name, baggageItem.Value);
+                            if (!requestTelemetry.Context.Properties.ContainsKey(baggageItem.Name))
+                            {
+                                requestTelemetry.Context.Properties[baggageItem.Name] = baggageItem.Value;
+                            }
+                        }
+                    }
+                }
+            }
+            else if (httpContext.Request.Headers.TryGetValue(RequestResponseHeaders.StandardRootIdHeader, out standardRootId))
+            {
+                activity.SetParentId(standardRootId);
+
+                StringValues standardParentId;
+                if (httpContext.Request.Headers.TryGetValue(RequestResponseHeaders.StandardParentIdHeader, out standardParentId))
+                {
+                    requestTelemetry.Context.Operation.ParentId = standardParentId;
+                }
+            }
+            activity.Start();
+            httpContext.Features.Set(activity);
+
+            requestTelemetry.Id = activity.Id;
+            requestTelemetry.Context.Operation.Id = activity.RootId;
+
+            this.client.Initialize(requestTelemetry);
+            requestTelemetry.Start(timestamp);
+            httpContext.Features.Set(requestTelemetry);
+
+            return requestTelemetry;
+        }
+
+        private void SetAppIdInResponseHeader(HttpContext httpContext, RequestTelemetry requestTelemetry)
+        {
+            IHeaderDictionary responseHeaders = httpContext.Response?.Headers;
+            if (responseHeaders != null &&
+                !string.IsNullOrEmpty(requestTelemetry.Context.InstrumentationKey) &&
+                (!responseHeaders.ContainsKey(RequestResponseHeaders.RequestContextHeader) || HttpHeadersUtilities.ContainsRequestContextKeyValue(responseHeaders, RequestResponseHeaders.RequestContextTargetKey)))
+            {
+                string correlationId = null;
+                if (this.correlationIdLookupHelper.TryGetXComponentCorrelationId(requestTelemetry.Context.InstrumentationKey, out correlationId))
+                {
```

Comment:
OperationCorrelationTelemetryInitializer also queries correlatioIdLookupHelper for the same instrumentation key.
I wonder if code from OperationCorrelationTelemetryInitializer could go here and then we can just get rid of it?